internal/controller/ols_app_server.go

```diff
@@ -0,0 +1,377 @@
+	}
+
+	service := &corev1.Service{}
+	err = r.Client.Get(ctx, client.ObjectKey{Name: OLSAppServerServiceName, Namespace: OLSNamespaceDefault}, service)
+	if err != nil && errors.IsNotFound(err) {
+		err = r.generateService(olsconfig, service)
+		if err != nil {
+			return fmt.Errorf("failed to generate OLS service: %w", err)
+		}
+		err = r.Create(ctx, service)
+		if err != nil {
+			return fmt.Errorf("failed to create OLS service: %w", err)
+		}
+	} else if err != nil {
+		return err
+	}
+	r.logger.Info("OLS service reconciled", "service", service.Name)
+
+	r.logger.Info("reconcileAppServer completes")
+
+	return nil
+}
+
+func (r *OLSConfigReconciler) generateServiceAccount(cr *olsv1alpha1.OLSConfig, sa *corev1.ServiceAccount) error {
+	*sa = corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      OLSAppServerServiceAccountName,
+			Namespace: OLSNamespaceDefault,
+		},
+	}
+
+	return nil
+}
+
+func (r *OLSConfigReconciler) generateOLSConfigMap(cr *olsv1alpha1.OLSConfig, cm *corev1.ConfigMap) error {
+
+	providerConfigs := []ProviderConfig{}
+	for _, provider := range cr.Spec.LLMConfig.Providers {
+		credentialPath := path.Join(APIKeyMountRoot, provider.CredentialsSecretRef.Name, BamApiKeyFileName)
+		if strings.ToLower(provider.Name) == "openai" {
+			credentialPath = path.Join(APIKeyMountRoot, provider.CredentialsSecretRef.Name, OpenaiApiKeyFileName)
+		}
+
+		modelConfigs := []ModelConfig{}
+		for _, model := range provider.Models {
+			modelConfig := ModelConfig{
+				Name: model.Name,
+				URL:  model.URL,
+			}
+			modelConfigs = append(modelConfigs, modelConfig)
+		}
+
+		providerConfig := ProviderConfig{
+			Name:            provider.Name,
+			URL:             provider.URL,
+			CredentialsPath: credentialPath,
+			Models:          modelConfigs,
+		}
+		providerConfigs = append(providerConfigs, providerConfig)
+	}
+
+	conversationCache := ConversationCacheConfig{
+		Type:   string(cr.Spec.OLSConfig.ConversationCache.Type),
+		Memory: MemoryCacheConfig{MaxEntries: cr.Spec.OLSConfig.ConversationCache.Memory.MaxEntries},
+	}
+
+	olsConfig := OLSConfig{
+		DefaultModel:    cr.Spec.OLSConfig.DefaultModel,
+		DefaultProvider: cr.Spec.OLSConfig.DefaultProvider,
+		Logging: LoggingConfig{
+			AppLogLevel: cr.Spec.OLSConfig.LogLevel,
+			LibLogLevel: cr.Spec.OLSConfig.LogLevel,
+		},
+		ConversationCache: conversationCache,
+	}
+
+	appSrvConfigFile := AppSrvConfigFile{
+		LLMProviders: providerConfigs,
+		OLSConfig:    olsConfig,
+	}
+	configFileStr, err := yaml.Marshal(appSrvConfigFile)
+	if err != nil {
+		return fmt.Errorf("failed to generate OLS config file %w", err)
+	}
+	hashFunc := sha256.New()
+	_, err = hashFunc.Write(configFileStr)
+	if err != nil {
+		return fmt.Errorf("failed to generate OLS config file hash %w", err)
+	}
+	configFileHash := fmt.Sprintf("%x", hashFunc.Sum(nil))
```

Comment:
Move hash compute out to separate  function.

Reply:
will move to utils.go